errors.go

```diff
@@ -100,6 +105,21 @@ func (p *Error) Error() string {
 	return output.String()
 }
 
+// ErrorChain returns an extended error message which includes a complete causal chain.
+// When errors are marshalled certain information is lost (e.g. the 'cause').  This means if an error travels through
+// a number of services (and it's potentially augmented at each hop) that the core error message may be lost.
+func (p *Error) ErrorChain() string {
+	msg := p.Error()
+	chain := p.MessageChain
+
+	// Chances are the latest error in the chain is already part of the standard error message.
+	if len(chain) > 0 && strings.Contains(msg, chain[0]) {
+		chain = chain[1:]
+	}
```

Comment:
This begs the questions when this would every not be the case? What about a more simple implementation that just assumes this is true, and just prints the code followed by the full chain?

Reply:
This is a reasonable question and I don't have examples of where this wouldn't currently hold true off the top of my head.  Stylistically I like to bias towards defensive code (within reason) and feel this is an example of this.  

The (theoretical) benefit here is that something could change in the future with the way core error messages are rendered and this method would continue to work - we can't be in a position where we reference an out of bounds index or drop a potentially useful message in the chain.

Reply:
Broadly feeling is it's better to try and fail fast and loudly if something changes from our expectations. The current logic looks like it tries to carry on even if there is a currently unknown case that crops up - leading to confusing behaviour. Anyway, I'll leave it to you to decide in this instance.

I'm actually concerned there may be a bit of a bug here. If you call call Augment on a terrors.Error,

```
terr := terrors.Augment(terrors.InternalService("foo", "inner", nil), "outer", nil)
```

wouldn't you end up with a state like:


```
Error{
  Code: "internal_service.foo",
  Message: "inner",
  MessageChain: []string{"outer", "inner"},
  cause: // inner,
}
```

calling `terr.Error()` would return `internal_service.foo: outer: inner`, so when you call `terr.MessageChain` it would print `internal_service.foo: outer: inner: inner` - i.e. you get the `inner` message twice because it's message is in the `cause` field (so part of `terr.Error()`) and also in the `MessageChain` field.

Just running this through in my head so I might have made a mistake, but this could be worth a test case to prove I'm wrong (if I am).

Reply:
It should only be possible to have 1 item added to `MessageChain` for a single call to `Augment`.  See screenshot below where I threw a test together to confirm what the resulting string would look like:
![image](https://user-images.githubusercontent.com/35197667/216395995-533986a9-17d5-4c57-a485-f30054122351.png)

> it's better to try and fail fast and loudly if something changes from our expectations

I don't disagree with this in general.  I wonder if it's what we'd want in this case though, esp. since this code will be used when logging errors - if we hit an index out of bounds error and the process panicked we'd end up getting no log.

Reply:
Ah I misunderstood the implementation of Augment - I thought `MessageChain` would include the message of the _current_ error too.

But then I think the problem I've pointed out still exists if an error is Augment a second time. For example 

terr := terrors.Augment(terrors.Augment(terrors.InternalService("foo", "inner 0", nil), "inner 1", nil), "outer", nil)

You'd get an ErrorChain() like `internal_service.foo: outer: inner_1: inner_0: inner_0`.

Admittedly this would only happen if Augment is called > 1 times on the same error chain within the same service.

All untested, so I may be making a mistake with my interpretation of this.

Reply:
You're right about that!

Solving this for all potential cases feels like it would result in more complexity that it warranted, so I've side stepped it in a way that I think makes sense.  

I've removed the `ErrorChain()` method 😀 This leaves it up to higher level code to decide how to best output error details.  For the use case I have in mind (logging) this will mean that the log message will just be the standard error however I'll add the items from the `MessageChain` as attributes on the log - if there's some duplication between what's here and what's in the message that'll be fine.